M. Afif Afianto

afifafian opened the pull request:
.env:
SECRET=jwtSECRET

email: admin3email.com
password: 12345

Deployment link:
heroku: https://murmuring-wave-46445.herokuapp.com/
firebase: https://e-commerce-cms-fa8ee.web.app/

afifafian committed:
First Commit:
- Added API doc in file readme.md
- Finished test for user and product endpoints

nadia12 commented on server/middlewares/errorHandler.js:
```diff
@@ -0,0 +1,40 @@
+
+function errorHandler(err, req, res, next) {
+    if(err.name === "SequelizeValidationError") {
+        const errMessages = []
```
[OPTIONAL] Kalau ada waktu, nanti coba baca2 tentang es6 built in lain, seperti map, filter, reduce, dll.. https://medium.com/poka-techblog/simplify-your-javascript-use-map-reduce-and-filter-bd02c593cc2d

nadia12 commented on server/middlewares/errorHandler.js:
```diff
@@ -0,0 +1,40 @@
+"use strict"
+
+function errorHandler(err, req, res, next) {
```
aku baru ngeh, kondisi kamu berdasar `message,`, akan lebih baik kondisi berdasar `nama`, karena `message` itu rentan bisa berubah... sedangkan `nama` itu biasanya lebih konsisten dan jarang berubah.

afifafian replied:
siap mbak nanti coba dipelajarin lagi untuk ubah kondisinya

nadia12 commented on server/controllers/UserController.js:
```diff
@@ -0,0 +1,52 @@
+            where: {email}
+        })
+        .then(function(data){
+            if(!data) {
```
secara logic, memang sudah betul. Hanya, kalau bisa hindari nested if yang terlalu dalam. jadi bisa direfactor kodenya.

nadia12 replied:
Bisa direfactor seperti ini:

```
if (!data) {
  throw {
    name: "Validation_error",
    statusCode: 404,
    message: "Data not found!",
  };
} else if (data && data.role !== "admin") {
  throw {
    name: "Validation_error",
    statusCode: 401,
    message: "Not authorized user!",
  };
} else if (data && !bcrypt.compareSync(password, data.password)) {
  throw {
    name: "Validation_error",
    statusCode: 400,
    message: "Incorrect Email or Password!",
  };
} else {           
  const token = jwtSign({ id: data.id, email: data.email, role: data.role });
  return res.status(200).json({ access_token: token });
}
```

coba dulu yaa.. ini codenya belum ditesting

nadia12 commented on server/models/user.js:
```diff
@@ -0,0 +1,73 @@
+
+  User.addHook('beforeCreate', function(instance, option){
+    const salt = bcrypt.genSaltSync(saltRounds);
+    const hash = bcrypt.hashSync(instance.password, salt);
```
masukkan ke helper ya. bcryptnya

afifafian replied:
Oh iya mbak yang ini belum dihapus kelupaan maaf

nadia12 commented on server/routes/productRouter.js:
```diff
@@ -0,0 +1,13 @@
+const ProductController = require('../controllers/ProductController')
+const {authentication, authorization} = require('../middlewares/auth')
+
+routes.get('/', authentication, authorization, ProductController.show)
```
pakai` routes.use('authentication, authorization')` diatasnya juga bisa, jadi gak disematkan satu2

afifafian replied:
Oke mbak siap nanti diubah

afifafian committed:
Finished on Client Folder:
- Added .vue files in views and components folder
- Vue Router Implementation done
- Navigation Guard done
- 404 Not Found page included in views folder
- Validation on client done

afifafian committed:
Added Procfile in server folder

afifafian committed:
Added Configuration for Production in config.json

afifafian committed:
updated package.json

afifafian committed:
Finished Deployment setup for Firebase:
- firebase.json
- Url updated for API endpoints in store/index.js

afifafian committed:
updated url
